### PR TITLE
Print nested ternaries differently

### LINKED ISSUE #737: alternate ternary formatting
Here is how Prettier currently reformats nested ternaries:
````
const value = condition1
  ? value1
  : condition2
      ? value2
      : condition3
          ? value3
          : value4;
````
This isn't so bad when there are only two ternaries, but I sometimes have many more because I essentially use them as an alternative to a switch statement when each case only gets a single value.

Here is how I prefer to format this case:
````
const value =
  condition1 ? value1 :
  condition2 ? value2 :
  condition3 ? value3 :
  value4;
````
I like this because it clearly associates each condition with its corresponding value and the indentation doesn't increase regardless of the number of ternaries.

Perhaps some like the current formatting because it discourages use of nested ternaries, but I think my example shows that you can have any number without making the code hard to read.

Would you consider adding an option to support this alternate formatting? Maybe something like this? `--ternary=condval`

### PR DESCRIPTION
**Pull request by suchipi**
Changes formatting of switch-style nested ternaries to the suggestion in https://github.com/prettier/prettier/pull/4767#issuecomment-401764876.

Fixes #737

- [x] I’ve added tests to confirm my change works.
- [ ] (n/a) (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: tests/ternaries/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -718,10 +718,10 @@ a
...
     ? bbbbbbbbbbbbbbbbbb
     : ccccccccccccccc
-        ? ddddddddddddddd
-        : eeeeeeeeeeeeeee
-            ? fffffffffffffff
-            : gggggggggggggggg;
+      ? ddddddddddddddd
+      : eeeeeeeeeeeeeee
+        ? fffffffffffffff
+        : gggggggggggggggg;
```
Comment (ikatyang) [inline tests/ternaries/__snapshots__/jsfmt.spec.js.snap]: This one seems not working.

File: tests/ternaries/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -1633,8 +2301,8 @@ a => a ? aasdasdasdasdasdasdaaasdasdasdasdasdasdasdasdasdasdasdasdasdaaaaaa : a
 debug ? (this.state.isVisible ? "partially visible" : "hidden") : null;
 debug
 	? this.state.isVisible && somethingComplex
-		? "partially visible"
-		: "hidden"
+	  ? "partially visible"
+	  : "hidden"
```
Comment (ikatyang) [inline tests/ternaries/__snapshots__/jsfmt.spec.js.snap]: This shouldn't be changed, we have a special case for ternary in ternary to use tabs in `--use-tabs` mode, that's what the removed part in 3018dd1ff5c2669dff0b3feb593f4ca9ba526175 wanted to achieve.

Comment (suchipi) [reply tests/ternaries/__snapshots__/jsfmt.spec.js.snap]: Ok, thanks. I will fix it

**Comment by lydell** [comment]
@suchipi Do you think you’ll be able to finish this one up?

**Comment by duailibe** [comment]
@ikatyang Can you check if 23cc0cc fixed it?

**Comment by ikatyang** [comment]
`--use-tabs` is working, but spaces are not.

**Prettier pr-5039**
[Playground link](https://deploy-preview-5039--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuEBDABAfnQRiQJkx0wGYkAWAkAGhAgAcYBLaAZ2VFQCcuIB3AArcE7FKgBuEJgBMaIAEZdUYANZwYAZXrKmUAObIYXAK5xaACxgBbADYB1c03ittYOBpFOm4pwE9k4KzstLqscFwwAkp6VqjIAGaoNmG0AFasAB4AQkqq6hqoVnAAMrpwCUkpINpcYVwB8qjyvjbQcvRcujB2MjDmyDi0HRBhdkr0AR1wdeLltFxwAI7GTAtRqDFxSInJZiBhVkyGJnusuno2cACKxhDwFbu0ME090n3IAKxPSkw25wDCECssQCUGgcxAxjCABUmqIdmEAL6IoA)
```sh
--parser babylon
--print-width 1
--tab-width 5
```

**Input:**
```jsx
a ? 1:2?1?3:4:2
```

**Output:**
```jsx
a
     ? 1
     : 2
     ? 1
       ? 3
       : 4
     : 2;

```

**Expected Output**
```jsx
a
     ? 1
     : 2
     ? 1
          ? 3
          : 4
     : 2;

```

File: tests/typescript_conditional_types/__snapshots__/jsfmt.spec.js.snap [hunk 3]
```diff
@@ -25,6 +33,16 @@ type DeepReadonlyObject<T> = {
...
 
+type TypeName<T> = T extends string
+  ? "string"
+  : T extends number
+  ? "number"
+  : T extends boolean
+  ? "boolean"
+  : T extends undefined
+  ? "undefined"
+  : T extends Function ? "function" : "object";
```
Comment (ikatyang) [inline tests/typescript_conditional_types/__snapshots__/jsfmt.spec.js.snap]: Unrelated to this PR: We should force breaking nested TS type ternaries just like normal ternaries.

**Comment by suchipi** [comment]
Thanks @ikatyang, my life has been really busy lately because I am getting a new job and my apartment got flooded, I really appreciate the help ❤️

**Comment by duailibe** [comment]
Great work @suchipi and @ikatyang. Thank you!

**Comment by davidmccabe** [comment]
This is a huge improvement, thanks for working on it.

Looking at the test file, it seems that one of the examples is still not intended, lines 5 – 11:

```
const value = condition1
? value1
: condition2
    ? value2
    : condition3
        ? value3
        : value4;
```

It's not clear why this example is formatted differently from the others. Is this intended?

**Comment by suchipi** [comment]
@davidmccabe could you link the the file and line that you're looking at?

It's possible that you're looking at the unformatted input. Our snapshots look like this:

```
input
~~~~~~~~~
output
```

**Comment by fbartho** [comment]
@suchipi / @ikatyang -- It looks like this PR implemented changes that led to this issue being opened: https://github.com/prettier/prettier/issues/5257 -- would somebody mind commenting in there to see if that issue should be instead marked as a "Won't-fix"/"Working as intended"?

If this is the new way, and that issue isn't a bug so much as a stylistic preference that doesn't match with the changes introduced here, then I'll just deal.

**Comment by lydell** [comment]
@fbartho This PR intended to flatten the “tail” of ternaries – if it caused #5257 it is a bug.

**Pull request merged**